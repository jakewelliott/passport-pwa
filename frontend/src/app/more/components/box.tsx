import { cn } from '@/lib/cn-helper';

interface BoxProps {
    title: string;
    headerClass: string;
    bodyClass: string;
    children: React.ReactNode;
}

export const Box = ({ title, headerClass, bodyClass, children }: BoxProps) => {
    return (
        <div className='flex-col'>
            <div
                className={cn(
                    'flex w-full items-center justify-center p-2 pt-3 text-center text-system_white',
                    headerClass,
                )}
            >
                <h3>{title}</h3>
            </div>
            <div className={cn('p-4', bodyClass)}>{children}</div>
        </div>
    );
};
