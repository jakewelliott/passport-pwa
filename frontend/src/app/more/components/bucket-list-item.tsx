import { useParks } from '@/hooks/queries/useParks';
import { FaRegCheckSquare, FaRegSquare } from 'react-icons/fa';
import DateHelper from '@/lib/date-helper';
import { useLocation } from '@/hooks/useLocation';
import { useUpdateBucketListItem } from '@/hooks/queries/useBucketList';
import { CompletedBucketListItem, BucketListItemInterface } from '@/lib/mock/types';

interface BucketListItemProps {
  bucketListItem: BucketListItemInterface;
  completedItems: CompletedBucketListItem[];
}

export const BucketListItem = ({ bucketListItem, completedItems }: BucketListItemProps) => {
  const isCompleted = completedItems.some(item => item.id === bucketListItem.id);
  const { data: parks } = useParks();
  const park = parks?.find(park => park.id === bucketListItem.park);
  const useUpdateBucketListItemMutation = useUpdateBucketListItem();
  const userLocation = useLocation();

  function updateBucketListItem(id: number, status: string): void {
    const useUpdateBucketListItemMutation = useUpdateBucketListItem(id);
    useUpdateBucketListItemMutation.mutate({
      latitude: userLocation.geopoint?.latitude ?? 0, 
      longitude: userLocation.geopoint?.longitude ?? 0, 
      status: status, 
      updatedAt: new Date()
    }, {
      onSuccess: () => {
        if (status === 'completed') {
          completedItems.push({
            id: bucketListItem.id,
            park: bucketListItem.park,
            updatedAt: DateHelper.stringify(new Date())
          });
        } else {
          completedItems = completedItems.filter(item => item.id !== bucketListItem.id);
        }
      }
    });
  }

  return (
    <div className='my-2.5 flex items-start'>
      {isCompleted ? (
        <FaRegCheckSquare
          data-testid='checked-icon'
          size={'24px'}
          strokeWidth={3}
          style={{ paddingRight: '5px', paddingTop: '3px' }}
          onClick={() => updateBucketListItem(bucketListItem.id, 'not completed')}
        />
      ) : (
        <FaRegSquare
          data-testid='unchecked-icon'
          size={'24px'}
          strokeWidth={3}
          style={{ paddingRight: '5px', paddingTop: '3px' }}
          onClick={() => updateBucketListItem(bucketListItem.id, 'completed')}
        />
      )}
      <div className='flex w-full flex-col justify-center'>
        <p>{bucketListItem.task}</p>
        <div className='flex flex-wrap items-center justify-between gap-2'>
          <p className='p-mini text-main_green'>{park?.addresses[0].city ?? ''}</p>
          {isCompleted && (
            <p className='p-mini'>{DateHelper.toStringShort(new Date(completedItems.find(item => item.id === bucketListItem.id)?.updatedAt ?? ''))}</p>
          )}
        </div>
      </div>
    </div>
  );
};
