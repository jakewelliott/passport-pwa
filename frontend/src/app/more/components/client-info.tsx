import DateHelper from '@/lib/date-helper';

export const ClientInfo = () => {
    // TODO: there's definitely a better way to do this
    const appVersion = '1.0.0';
    // TODO: this doesn't work for some reason
    const connectionStatus = window.navigator.onLine ? 'Online' : 'Offline';
    // TODO: implement this
    const lastCached = DateHelper.toStringLong(new Date());

    return (
        <div className='grid w-full grid-cols-2'>
            <p>App Version:</p>
            <p className='text-right'>{appVersion}</p>
            <p>Connection Status:</p>
            <p className='text-right'>{connectionStatus}</p>
            <p>Last Cached:</p>
            <p className='text-right'>{lastCached}</p>
        </div>
    );
};
