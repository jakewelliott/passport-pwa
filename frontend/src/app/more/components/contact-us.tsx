import { Link } from 'react-router-dom';
import { FaFacebookF, FaYoutube, FaInstagram } from 'react-icons/fa';

export const ContactUs = () => {
  return (
    <div className='p-4'>
      <div className='bg-supporting_inactiveblue p-2 text-center text-system_white'>
        <h3>CONTACT US</h3>
      </div>
      <div className='space-y-2 bg-supporting_lightblue p-2.5 text-center'>
        <div>
          <Link to={'www.ncparks.gov'} className='text-supporting_inactiveblue no-underline'>
            www.ncparks.gov
          </Link>
        </div>
        <div className='flex justify-center gap-3'>
          <Link
            to={'https://www.facebook.com/NorthCarolinaStateParks/'}
            className='flex h-8 w-8 items-center justify-center rounded-full bg-supporting_inactiveblue'
          >
            <FaFacebookF size={20} className='text-supporting_lightblue' />
          </Link>
          <Link
            to={'https://www.instagram.com/ncstateparks/'}
            className='flex h-8 w-8 items-center justify-center rounded-full bg-supporting_inactiveblue'
          >
            <FaInstagram size={20} className='text-supporting_lightblue' />
          </Link>
          <Link
            to={'https://www.youtube.com/@parksNC'}
            className='flex h-8 w-8 items-center justify-center rounded-full bg-supporting_inactiveblue'
          >
            <FaYoutube size={20} className='text-supporting_lightblue' />
          </Link>
        </div>
        <p className='space-y-1'>
          <span>NC Division of Parks and Recreation</span>
          <br />
          <span>Dept. of Natural and Cultural Resources</span>
          <br />
          <span>1615 Mail Service Center</span>
          <br />
          <span>Raleigh, NC 27699</span>
          <br />
          <span>(919) 707-9300</span>
        </p>
      </div>
    </div>
  );
};
