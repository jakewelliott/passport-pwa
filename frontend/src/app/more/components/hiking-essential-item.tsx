import type React from 'react';

interface HikingEssentialItemProps {
    imageSrc: string;
    altText: string;
    title: string;
    description: string;
}
