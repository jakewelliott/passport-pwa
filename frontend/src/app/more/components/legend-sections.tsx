import { BLAZE_ICONS, BLUE_ICONS, GREEN_ICONS, type PARK_ICONS, type ParkIcon, RED_ICONS } from '@/types/icons';

type IconWithExtraText = ParkIcon;

interface IconSection {
    sectionName: string;
    icons: IconWithExtraText[];
}

type IconName = (typeof PARK_ICONS)[number]['iconName'];

const textHelper = (icon: ParkIcon) => ({
    ...icon,
});

// @biome-ignore lint/style/noNonNullAssertion: we know this icon exists
const ICON_SECTIONS: IconSection[] = [
    {
        sectionName: 'Camping',
        icons: GREEN_ICONS.map(textHelper),
    },
    {
        sectionName: 'Activities',
        icons: RED_ICONS.map(textHelper),
    },
    {
        sectionName: 'Amenities',
        icons: BLUE_ICONS.map(textHelper),
    },
    {
        sectionName: 'State Trail Stamps',
        icons: BLAZE_ICONS.map(textHelper),
    },
];

const path = (iconName: IconName) => `/icons/park/${iconName}.svg`;

const IconView = ({ icon }: { icon: IconSection['icons'][number] }) => {
    return (
        <div key={icon.iconName} className='my-1 flex items-center'>
            <img src={path(icon.iconName)} alt={icon.tooltip} width={'36px'} height={'36px'} />
            <div className='ml-2 flex w-full flex-col justify-center'>
                <p>{icon.tooltip}</p>
                {icon.extraText && <p className='max-w-56 p-mini'>*{icon.extraText}</p>}
            </div>
        </div>
    );
};

export const LegendSections = () => {
    return (
        <>
            {ICON_SECTIONS.map((section) => (
                <div className='m-4' key={section.sectionName}>
                    <h4 className={`pb-3 text-icon_${section.sectionName.toLowerCase()}`}>{section.sectionName}</h4>
                    {section.icons.map((icon) => (
                        <IconView key={icon.iconName} icon={icon} />
                    ))}
                </div>
            ))}
        </>
    );
};
