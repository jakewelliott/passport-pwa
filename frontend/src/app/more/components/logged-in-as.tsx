import { useUser } from '@/hooks/queries/useUser';
import { useNavigate } from 'react-router-dom';

export const LoggedInAs = () => {
	const { data: user } = useUser();
	const navigate = useNavigate();


	const message = user?.username ? `You are currently logged in as '${user.username}'` : 'You are not logged in';

	return (
		<div className='text-center'>
			{message}
			<br />
			<button className='link' onClick={() => navigate('/logout')} type='button'>
				Log out
			</button>
		</div>
	);
};
