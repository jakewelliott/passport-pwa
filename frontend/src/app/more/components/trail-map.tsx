export const TrailMap = () => {
    return (
        <div className='flex flex-col'>
            <img src={'/PassportMap_TrailsOnly.svg'} alt='NC State Map showing all trails' />
            <div className='ml-2 flex w-full justify-around'>
                <div>
                    <p className='text-trail-danriver'>Dan River</p>
                    <p className='text-trail-deepriver'>Deep River</p>
                    <p className='text-trail-eastcoastgreenway'>East Coast Greenway</p>
                    <p className='text-trail-equine'>Equine</p>
                    <p className='text-trail-fontaflora'>Fonta Flora</p>
                    <p className='text-trail-frenchbroadriver'>French Broad River</p>
                    <p className='text-trail-hawriver'>Haw River</p>
                </div>
                <div>
                    <p className='text-trail-hickorynutgorge'>Hickory Nut Gorge</p>
                    <p className='text-trail-mountainstosea'>Mountains-to-Sea</p>
                    <p className='text-trail-northernpeaks'>Northern Peaks</p>
                    <p className='text-trail-overmountainvictory'>Overmountain Victory</p>
                    <p className='text-trail-roanokeriver'>Roanoke River</p>
                    <p className='text-trail-wildernessgateway'>Wilderness Gateway</p>
                    <p className='text-trail-yadkinriver'>Yadkin River</p>
                </div>
            </div>
        </div>
    );
};
