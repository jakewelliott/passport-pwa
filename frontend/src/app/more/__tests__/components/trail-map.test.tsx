import { TrailMap } from '@/app/more/components/trail-map';
import { render, screen } from '@testing-library/react';

describe('TrailMap', () => {
  it('renders the map image with correct attributes', () => {
    render(<TrailMap />);

    const mapImage = screen.getByAltText('NC State Map showing all trails');
    expect(mapImage).toBeInTheDocument();
    expect(mapImage).toHaveAttribute('src', '/PassportMap_TrailsOnly.svg');
  });

  it('renders all trail names with correct styling', () => {
    render(<TrailMap />);

    const trailsAndClasses = {
      'Dan River': 'text-trail_danriver',
      'Deep River': 'text-trail_deepriver',
      'East Coast Greenway': 'text-trail_eastcoastgreenway',
      Equine: 'text-trail_equine',
      'Fonta Flora': 'text-trail_fontaflora',
      'French Broad River': 'text-trail_frenchbroadriver',
      'Haw River': 'text-trail_hawriver',
      'Hickory Nut Gorge': 'text-trail_hickorynutgorge',
      'Mountains-to-Sea': 'text-trail_mountainstosea',
      'Northern Peaks': 'text-trail_northernpeaks',
      'Overmountain Victory': 'text-trail_overmountainvictory',
      'Roanoke River': 'text-trail_roanokeriver',
      'Wilderness Gateway': 'text-trail_wildernessgateway',
      'Yadkin River': 'text-trail_yadkinriver',
    };

    for (const [trailName, className] of Object.entries(trailsAndClasses)) {
      const trailElement = screen.getByText(trailName);
      expect(trailElement).toBeInTheDocument();
      expect(trailElement).toHaveClass(className);
    }
  });

  it('has correct layout structure', () => {
    render(<TrailMap />);

    const container = screen.getByAltText('NC State Map showing all trails').closest('div');
    const legendContainer = screen.getByText('Dan River').closest('div')?.parentElement;

    expect(container).toHaveClass('flex', 'flex-col');
    expect(legendContainer).toHaveClass('ml-2', 'flex', 'w-full', 'justify-around');
  });

  it('splits trail names into two columns', () => {
    render(<TrailMap />);

    const firstColumnTrails = [
      'Dan River',
      'Deep River',
      'East Coast Greenway',
      'Equine',
      'Fonta Flora',
      'French Broad River',
      'Haw River',
    ];
    const secondColumnTrails = [
      'Hickory Nut Gorge',
      'Mountains-to-Sea',
      'Northern Peaks',
      'Overmountain Victory',
      'Roanoke River',
      'Wilderness Gateway',
      'Yadkin River',
    ];

    const firstColumn = screen.getByText('Dan River').closest('div');
    const secondColumn = screen.getByText('Hickory Nut Gorge').closest('div');

    for (const trail of firstColumnTrails) {
      expect(firstColumn).toContainElement(screen.getByText(trail));
    }

    for (const trail of secondColumnTrails) {
      expect(secondColumn).toContainElement(screen.getByText(trail));
    }
  });
});
