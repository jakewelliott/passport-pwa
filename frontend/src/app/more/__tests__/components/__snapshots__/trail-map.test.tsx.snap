// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`TrailMap > matches snapshot 1`] = `
<div>
  <div
    class="flex flex-col"
  >
    <img
      alt="NC State Map showing all trails"
      src="/PassportMap_TrailsOnly.svg"
    />
    <div
      class="ml-2 flex w-full justify-around"
    >
      <div>
        <p
          class="text-trail-danriver"
        >
          Dan River
        </p>
        <p
          class="text-trail-deepriver"
        >
          Deep River
        </p>
        <p
          class="text-trail-eastcoastgreenway"
        >
          East Coast Greenway
        </p>
        <p
          class="text-trail-equine"
        >
          Equine
        </p>
        <p
          class="text-trail-fontaflora"
        >
          Fonta Flora
        </p>
        <p
          class="text-trail-frenchbroadriver"
        >
          French Broad River
        </p>
        <p
          class="text-trail-hawriver"
        >
          Haw River
        </p>
      </div>
      <div>
        <p
          class="text-trail-hickorynutgorge"
        >
          Hickory Nut Gorge
        </p>
        <p
          class="text-trail-mountainstosea"
        >
          Mountains-to-Sea
        </p>
        <p
          class="text-trail-northernpeaks"
        >
          Northern Peaks
        </p>
        <p
          class="text-trail-overmountainvictory"
        >
          Overmountain Victory
        </p>
        <p
          class="text-trail-roanokeriver"
        >
          Roanoke River
        </p>
        <p
          class="text-trail-wildernessgateway"
        >
          Wilderness Gateway
        </p>
        <p
          class="text-trail-yadkinriver"
        >
          Yadkin River
        </p>
      </div>
    </div>
  </div>
</div>
`;
